modules/call-dispatch/src/lib.rs

```diff
@@ -0,0 +1,348 @@
+	/// event with this id + dispatch result.
+	type MessageId: Parameter;
+	/// The overarching dispatch call type.
+	type Call: Parameter + GetDispatchInfo + Dispatchable<Origin = <Self as frame_system::Trait>::Origin>;
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as Trait>::MessageOrigin,
+		<T as Trait>::MessageId,
+	{
+		/// Message has been rejected by dispatcher because of spec version mismatch.
+		MessageVersionSpecMismatch(MessageOrigin, MessageId),
+		/// Message has been rejected by dispatcher because of weight mismatch.
+		MessageWeightMismatch(MessageOrigin, MessageId),
+		/// Message has been dispatched with given result.
+		MessageDispatched(MessageOrigin, MessageId, DispatchResult),
+	}
+);
+
+decl_module! {
+	/// Call Dispatch FRAME Pallet.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Deposit one of this module's events by using the default implementation.
+		fn deposit_event() = default;
+	}
+}
+
+impl<T: Trait> MessageDispatch<T::MessageOrigin, T::MessageId> for Module<T>
+where
+	<<T as Trait>::Call as Dispatchable>::PostInfo: sp_std::fmt::Debug,
+{
+	type Message = (SpecVersion, Weight, <T as Trait>::Call);
+
+	fn dispatch(origin: T::MessageOrigin, id: T::MessageId, message: Self::Message) {
+		let (spec_version, weight, call) = message;
+
+		// verify spec version
+		// (we want it to be the same, because otherwise we may decode Call improperly)
+		let expected_version = <T as frame_system::Trait>::Version::get().spec_version;
+		if spec_version != expected_version {
+			frame_support::debug::trace!(
+				"Message {:?}/{:?}: spec_version mismatch. Expected {:?}, got {:?}",
+				origin,
+				id,
+				expected_version,
+				spec_version,
+			);
+			Self::deposit_event(Event::<T>::MessageVersionSpecMismatch(origin, id));
+			return;
+		}
+
+		// verify weight
+		// (we want passed weight to be at least equal to pre-dispatch weight of the call
+		// because otherwise Calls may be dispatched at lower price)
+		let expected_weight = call.get_dispatch_info().weight;
+		if weight < expected_weight {
+			frame_support::debug::trace!(
+				"Message {:?}/{:?}: passed weight is too low. Expected at least {:?}, got {:?}",
+				origin,
+				id,
+				expected_weight,
+				weight,
+			);
+			Self::deposit_event(Event::<T>::MessageWeightMismatch(origin, id));
+			return;
+		}
+
+		// finally dispatch message
+		let origin_account = Self::bridge_account_id(&origin);
+		let dispatch_result = call.dispatch(frame_system::RawOrigin::Signed(origin_account).into());
+		frame_support::debug::trace!(
+			"Message {:?}/{:?} has been dispatched. Result: {:?}",
+			origin,
+			id,
+			dispatch_result,
+		);
+
+		Self::deposit_event(Event::<T>::MessageDispatched(
+			origin,
+			id,
+			dispatch_result.map(drop).map_err(|e| e.error),
+		));
+	}
+}
+
+impl<T: Trait> Module<T> {
+	fn bridge_account_id(origin: &T::MessageOrigin) -> T::AccountId {
```

Comment:
Just realised that it might be useful in other pallets, so should probably be part of some `bp-` crate (maybe same place as `InstanceId`).

Reply:
Extracted. Also removed recently introduced `MessageOrigin` - message origin is always `InstanceId` now.